### ISSUE TITLE
some questions about your code

### ISSUE THREAD

**Issue by xiaowangba2dai** [BrizziB/Graph-Classification-with-GCN#1]
Hello, i want to know the meaning about the "tol" and the "num_feats" in your code. It shows that "tol =num_nodes + num_graphs", but i don't know why to do this. Can you explain it for me, thank you!

**Comment by BrizziB** [comment]
Hi,  thank you for your question!
"num_feats" represents the number of features of each node, i.e. how many attributes each node has.
"num_nodes" represents how many nodes i have in all graphs,
"num_graphs" is the number of graphs (in "PROTEINS" dataset there are 43471 nodes distributed on 1113 graphs; each graph representing some king of protein i think)

In this piece of code:  
 `    if not FLAGS.with_pooling:
        input_dim = tot
    else:
        input_dim = num_nodes`

you can see that, if the chosen method is NOT "with_pooling" , then  you set the dimension of the adjacency matrix to "tot"="num_nodes"+"num_graphs";
this is because the "GlobalNodes" method requires a set of extra-nodes (one for each graph) in order to work.
Each "extra node" must be linked to each node of the corresponding graph.
In this way, that extra node will aggregate information from every node of the graph.

[https://github.com/tkipf/gcn/issues/4](url) here you will find further informations on this approach

**Comment by xiaowangba2dai** [comment]
Thank you for your answer! I try to build my own dataset, but how can i get the file "dataset_name_node_attributes.txt"?

**Comment by BrizziB** [comment]
That file, or an equivalent one, should be provided by your dataset; 
To fully exploit the benefits of GCN you should use dataset with node attributes,
however you can try to learn node embeddings even with no provided node attributes, do that by setting the "Featureless" flag to true; **I didn't test that part of the code very well, so i cannot affirm that it will work correctly (!!)**

**Comment by AmirSh15** [comment]
Hi @BrizziB ,
I want to train on my dataset. However mine has different adjacency matrix. I mean my dataset has 1000 graphs and each graph has a related adjacency matrix and features.
How can I changed your code to be fed with different adjacency matrices?

**Comment by BrizziB** [comment]
Hi @AmirSh15 ,
you can use this version of gcn without great changes:
You have to build a unique adjacency matrix; this matrix must contain all your adjacency matrices in a block diagonal form. 
You can refer to this https://github.com/tkipf/gcn/issues/4 to get further help, let me know if you succeed or have problems

**Comment by IMSUSHI** [comment]
Hi @BrizziB ,
Thank you for your great code. I am wondering how to build the block-diagonal matrix when my input graphs have different sizes. Does it need the unique adjacency matrix you mentioned above?

**Comment by BrizziB** [comment]
Hi,
yes, you should definitely build that adjacency matrix as illustrated in the reference link of my previous comment. The different size of single adjacency matrices is not an issue, you will just have different block size on the diagonal of the complete matrix

**Comment by IMSUSHI** [comment]
Hi @BrizziB ,
Many thanks for your reply. I've run my graph using both the "global node" approach and the global mean pooling approach, but unfortunately neither of them got a good result (the loss and accuracy value didn't converge at last). Why did this happened? Is it because of the small amount of the training data set? (I only have thirty graphs and 120 nodes for training and testing). Looking forward for your answer.
[BUILDINGS_A.txt](https://github.com/BrizziB/Graph-Classification-with-GCN/files/3599655/BUILDINGS_A.txt)
[BUILDINGS_graph_indicator.txt](https://github.com/BrizziB/Graph-Classification-with-GCN/files/3599656/BUILDINGS_graph_indicator.txt)
[BUILDINGS_graph_labels.txt](https://github.com/BrizziB/Graph-Classification-with-GCN/files/3599657/BUILDINGS_graph_labels.txt)
[BUILDINGS_node_attributes.txt](https://github.com/BrizziB/Graph-Classification-with-GCN/files/3599658/BUILDINGS_node_attributes.txt)
[PROTEINS_node_labels.txt](https://github.com/BrizziB/Graph-Classification-with-GCN/files/3599660/PROTEINS_node_labels.txt)
![the loss value of the global node approach](https://user-images.githubusercontent.com/54923691/64679294-0a142080-d4ae-11e9-8492-48fac94e8c4d.PNG)

**Comment by IMSUSHI** [comment]
Hi @BrizziB ,
Many thanks for your reply. I've run my graph using both the "global node" approach and the global mean pooling approach, but unfortunately neither of them got a good result (the loss and accuracy value didn't converge at last). Why did this happened? Is it because of the small amount of the training data set? (I only have thirty graphs and 120 nodes for training and testing). Looking forward for your answer.
Here's the data of my experiment:
[BUILDINGS_graph_indicator.txt](https://github.com/BrizziB/Graph-Classification-with-GCN/files/3599673/BUILDINGS_graph_indicator.txt)
[BUILDINGS_graph_labels.txt](https://github.com/BrizziB/Graph-Classification-with-GCN/files/3599674/BUILDINGS_graph_labels.txt)
[BUILDINGS_node_attributes.txt](https://github.com/BrizziB/Graph-Classification-with-GCN/files/3599675/BUILDINGS_node_attributes.txt)
[BUILDINGS_A.txt](https://github.com/BrizziB/Graph-Classification-with-GCN/files/3599672/BUILDINGS_A.txt)
![the loss value of the global node approach](https://user-images.githubusercontent.com/54923691/64679499-6d05b780-d4ae-11e9-96db-a0ef475be0e7.PNG)

**Comment by IMSUSHI** [comment]
Hi @BrizziB ,
Many thanks for your reply. I've run my graph using both the "global node" approach and the global mean pooling approach, but unfortunately neither of them got a good result (the loss and accuracy value didn't converge at last). Why did this happened? Is it because of the small amount of the training data set? (I only have thirty graphs and 120 nodes for training and testing). Looking forward for your reply.
Here's the data of my experiment:
[BUILDINGS_A.txt](https://github.com/BrizziB/Graph-Classification-with-GCN/files/3599678/BUILDINGS_A.txt)
[BUILDINGS_graph_indicator.txt](https://github.com/BrizziB/Graph-Classification-with-GCN/files/3599679/BUILDINGS_graph_indicator.txt)
[BUILDINGS_graph_labels.txt](https://github.com/BrizziB/Graph-Classification-with-GCN/files/3599680/BUILDINGS_graph_labels.txt)
[BUILDINGS_node_attributes.txt](https://github.com/BrizziB/Graph-Classification-with-GCN/files/3599681/BUILDINGS_node_attributes.txt)
![the loss value of the global node approach](https://user-images.githubusercontent.com/54923691/64679685-ca9a0400-d4ae-11e9-846e-4018174aeb6f.PNG)

**Comment by BrizziB** [comment]
As you suggested, it's probably because your dataset has too few graphs: the datasets used for with this methods usually have more than a thousand graphs.
Deep learning approaches usually needs lots of data, sadly.

**Comment by IMSUSHI** [comment]
Hi @BrizziB ,
Sorry to keep asking question. I added more graphs to my dataset, and got a good result. Now I'm wondering how to predict unlabled graphs using the parameters I got.
Looking forward for your reply.

**Comment by BrizziB** [comment]
You can save the weights you learned and reuse them later, that should be covered in a TensorFlow guide i think.
